src/streamed_orderbook.ts

```diff
@@ -0,0 +1,267 @@
+      }
+    }
+
+    return false
+  }
+
+  private applyEvents(events: EventData[]) {
+    this.markReverted(events)
+    for (const ev of events) {
+      const data = ev.returnValues
+      switch (ev.event) {
+        case 'Deposit':
+          this.updateBalance(
+            data.user,
+            data.token,
+            amount => amount + BigInt(data.amount),
+          )
+          break
+        case 'OrderCancellation':
+          this.account(data.owner).orders[parseInt(data.id)]!.validUntil = null
+          break
+        case 'OrderDeletion':
+          this.account(data.owner).orders[parseInt(data.id)] = null
+          break
+        case 'OrderPlacement':
+          this.account(data.owner).orders.push({
+            buyToken: data.buyToken,
+            sellToken: data.sellToken,
+            validFrom: parseInt(data.validFrom),
+            validUntil: parseInt(data.validUntil),
+            priceNumerator: BigInt(data.priceNumerator),
+            priceDenominator: BigInt(data.priceDenominator),
+            remainingAmount: BigInt(data.priceDenominator),
+          })
+          break
+        case 'SolutionSubmission':
+          if (!data.ext.reverted) {
+            this.updateBalance(
+              data.submitter,
+              '0',
+              amount => amount + BigInt(data.burntFees),
+            )
+            for (let i = 0; i < data.prices.length; i++) {
+              this.state.prices[data.tokenIdsForPrice[i]] = BigInt(data.prices[i])
+            }
+          }
+          break
+        case 'TokenListing':
+          this.state.tokens[data.id] = data.token
+          this.state.tokens[data.token] = data.id
+          break
+        case 'Trade':
+          if (!data.ext.reverted) {
+            this.updateBalance(
+              data.owner,
+              data.sellToken,
+              amount => amount - BigInt(data.executedSellAmount),
+            )
+            this.account(data.owner)
+              .orders[parseInt(data.orderId)]!
+              .remainingAmount -= BigInt(data.executedSellAmount)
+            this.updateBalance(
+              data.owner,
+              data.buyToken,
+              amount => amount + BigInt(data.executedBuyAmount),
+            )
+          }
+          break
+        case 'TradeReversion':
+          break
+        case 'WithdrawRequest':
+          this.updateWithdraw(
+            data.user,
+            data.token,
+            _ => BigInt(data.amount),
+          )
+          break
+        case 'Withdraw':
+          this.updateBalance(
+            data.user,
+            data.token,
+            amount => amount - BigInt(data.amount),
+          )
+          this.updateWithdraw(
+            data.user,
+            data.token,
+            _ => BigInt(0),
+          )
+          break
+        default:
+          console.error(ev)
+          throw new Error('unhandled event')
+      }
+    }
+  }
+
+  private markReverted(events: EventData[]) {
```

Comment:
nit: `mark revertedTrades` I would find it a little cleaner if this returned a Set of trades that have been reverted and we check inclusion rather than expecting some property is set to a magic value.

Reply:
This method is actually not needed. I removed it completely already as I was cleaning up the code as this can be done in a pretty straight forward way by just tracking the last solution and reverted trade events.

I also didn't really like this as it sneakily mutated the event data and felt a bit icky.